shell: backends: bt: Add Bluetooth LE

ubieda opened the pull request:
This PR implements the Bluetooth Backend for Shell, allowing to support Shell commands over BLE GATT.

Some technical details:
- **Endpoint:** a Custom GATT Service (NUS UUID), with two Characteristics to send and receive commands.
- **Commands:** sent to the device through a GATT Write command over RX Characteristic.
- **Responses:** returned from the device through GATT Notifications over TX Characteristic (provided the GATT client is subscribed).

**Pending Actions:**
- [x] Provide [User App sample](https://github.com/ubieda/bleak/blob/feature/interactive-terminal-over-bt/examples/uart_service.py) that exercises the Shell Backend.
- [x] Modify implementation to be compatible with NUS user-apps.
- [x] Add sample documentation.

ubieda committed:
shell: backends: bt: Add Bluetooth LE

Conveyed as a GATT Service, with two characteristics: one to write
commands through GATT write requests (on RX Characteristic), and the
other one to receive command responses through GATT Notifications (on
TX Characteristic).
The BT GATT UUIDs are compatible with NUS and should work with BT NUS
applications.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda commented:
Just a comment before going any further: I'm not sure if this PR should be preceeded by an Issue with RFC. Please give me a heads-up and I'll expand on the implementation details there.

Also, I'm up for presenting the design either in a technical meeting if it's of interest (Architecture or Bluetooth).

carlescufi commented:
> Just a comment before going any further: I'm not sure if this PR should be preceeded by an Issue with RFC. Please give me a heads-up and I'll expand on the design considerations there.
> 
> Also, I'm up for presenting the design either in a technical meeting if it's of interest (Architecture, Dev, Bluetooth or similar).

I think we can start with PR reviews first. If there's disagreements we can then discuss it in a meeting.

jori-nordic commented:
I just have a drive-by comment (haven't looked at diff):
We already have a log backend, now a shell backend.

Could we instead make a uart driver that goes over GATT? 
Nordic has the NUS protocol, and it's implemented in the log backend so we could reuse that. Bonus is it works well with the mobile apps.

(It's an old issue, but it checks out :p https://github.com/zephyrproject-rtos/zephyr/issues/34949)

ubieda commented:
I'd like to share more context on an application I'm using this for: We have Shell both over USB (uart-backend) and also BLE. The Shell over USB is used for factory provisioning and manufacturing testing, whereas BLE is mostly used for user interaction once deployed in the field. This is possible because of the nature of Shell backends, allowing to have more than one enabled simultaneously.

> Could we instead make a uart driver that goes over GATT?

@jori-nordic If I understood this correctly, by making the Shell over Bluetooth an instance of Shell over UART, we would not be able to have both: UART and BLE.

jori-nordic commented:
Yes, you're right. Having a different backend makes more sense.

Note that this is a limitation of the uart backend, there should be no technical reason the shell couldn't use multiple uart backends. We'd just have to move away from the DTS `chosen` prop and use an actual driver (e.g. `compatible = zephyr,shell-uart` or smth) directly on the `uart_x` nodes.

I'll review and test the PR. What app are you using to connect to the shell?

alwa-nordic commented:
Looks good mostly, but this is missing synchronization between the Bluetooth `connected`/`disconnected` event handlers and the `write` function. I think this is susceptible to race conditions.

ubieda commented:
> Looks good mostly, but this is missing synchronization between the Bluetooth connected/disconnected event handlers and the write function. I think these will run in different threads, so it's susceptible to race conditions.


@alwa-nordic Thanks for the feedback! I'll add some logic to sync the write function with connect/disconnect events.

ubieda commented:
> I'll review and test the PR. What app are you using to connect to the shell?

@jori-nordic Thanks! For internal testing I've used LightBlue with UTF8 encoding (see pic below). I can also put together a CLI script that acts as a client, if that makes it easier.
<img width="1247" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/65253049/850f2d3c-cfbd-4736-a5b7-e12e9470dbd6">

jori-nordic commented on subsys/shell/backends/shell_bt.c:
```diff
@@ -0,0 +1,228 @@
+
+	*cnt = length;
+
+	LOG_INF("%s: %d", __func__, length);
```
This should probably be `LOG_DBG`.

jori-nordic commented:
@ubieda I am completely unable to make lightblue communicate with it. I'm on android and I see you're on ios or mac, so the app + the host stack might be different. I can write but don't receive anything back, even though it shows me as subscribed.

nRF connect app slightly works, at least I'm able to get the "PS1" (ie `bt:~$ `), but not the whole output, even though I see the write fn being called w/ the appropriate lengths in the logs.

Seeing that it's not trivial to use, I would appreciate having a script in the tree along with the sample.
It could also be another sample (like `echo_bot`) that acts as client and just bridges the hardware UART to the GATT service.

ubieda commented:
> Seeing that it's not trivial to use, I would appreciate having a script in the tree along with the sample.
> It could also be another sample (like echo_bot) that acts as client and just bridges the hardware UART to the GATT service.

@jori-nordic sounds good! I'll incorporate either or both on this PR

alwa-nordic commented on subsys/shell/backends/shell_bt.c:
```diff
@@ -0,0 +1,228 @@
+		return;
+	}
+
+	shell_bt_st.bt.conn = conn;
```
Yes. It also needs to be behind a mutex.

jori-nordic commented:
A last feedback point, kinda nitpicky: we generally use the imperative form for commit messages. E.g. "Added foo function" -> "Add foo function". Not sure if there's an official guideline somewhere.

ubieda commented:
Made the following changes:
- Added logic to guard usage of the BT connection object, following @alwa-nordic suggestions.
- Modified commit and PR message to use imperative mood instead of past-tense.
- Addressed formatting and minor observations.

I'll now work towards having a BT client app (FW) to properly exercise this backend.

ubieda committed:
samples: shell: bt: Add sample for device with Shell over BLE

Device implements Shell with the Bluetooth LE backend, and starts
the sample by enabling BLE advertisements. Once connected, exposes NUS
Service to allow users sending commands and receiving responses.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

alwa-nordic commented on subsys/shell/backends/shell_bt.c:
```diff
@@ -0,0 +1,282 @@
+	LOG_HEXDUMP_DBG(data, length, "data_write");
+
+	/* It is not safe to use BT APIs with the conn mutex locked: rely on a reference copy */
+	(void)k_mutex_lock(shell_bt_st.bt.conn.mutex, K_FOREVER);
```
The mutex is taken inside `get_conn_ref_copy`, so it should not be taken here as well.

Side note: In fact, taking the mutex twice like this works only because the Zephyr mutex is 're-entrant'. Not all 'mutex' implementations are re-entrant. It it contested in the general CS community whether mutex re-entrancy is a desirable feature at all. I think it's awkward and more likely to hide bugs than be useful. :) /purist

ubieda replied:
Gotchya - Removed redundant mutex guard.

Let's keep us all happy if we can :smile:

nordic-krch commented:
> @jori-nordic If I understood this correctly, by making the Shell over Bluetooth an instance of Shell over UART, we would not be able to have both: UART and BLE.

Regarding having two instances of shell with UART. I think that it is supported. I was not aware of that but when I removed multi-instance shell UART support then Meta guys wanted to bring it back as they were using it (#66218). So having UART-over-BLE might be beneficial as it could be reused between logging and shell backend.

ubieda commented:
@nordic-krch Thanks for bringing this up! I was not aware that it could be possible.

Since we now agree there’s more than one way to implement the feature “Shell over BLE”, I’d like to capture some comments that could be useful to evaluate these alternative options (BT as a Shell Backend vs UART as a Shell Backend + UART-over-BLE driver):

- **Consistency with existing backends:** IMO, the abstraction of a new backend behind UART (BLE in this case), goes against the approach taken so far for this subsystem. As of today, there are dual/separate backend implementations between the Logging Subsystem and Shell (UART and RTT, at least), and I don’t see why BLE should be treated differently. I don’t know if the project has interest in moving away from the backends concept in pro of abstracting them behind common-code and gaining reusability between different subsystems. Please, let me know if that’s the case.
- **Accessibility for Application developers:** The way the Shell subsystem is structured today, makes it easy for app developers to understand what are the different backend options at their disposal. The devil is in the details, but I believe having BLE listed as another backend would be beneficial in terms of clarity and accessibility vs making it agnostic for the Shell subsystem.
- **Ease of Maintenance:** I agree in this respect, the abstraction of BLE behind UART makes it easier to maintain.
- **Compatibility with existing Client Apps:** This is a good observation, and I believe this could be achieved with both approaches. The main app that comes to my mind when thinking of Serial over BLE is NUS-compatible apps; as such, we can make the GATT database (Service and Characteristic) match the NUS so that it is agnostic from the app-clients.

Based on these observations, my suggestion are:
- Keep the BLE implementation as a separate Shell backend, and not abstract them behind UART.
- Make the BLE backend GATT database and operations compatible with existing NUS-compatible clients.

Please, LMK what are your thoughts @nordic-krch @jori-nordic

ubieda commented:
@jori-nordic sounds good!

I'll go ahead and make the changes to make this PR testable with NUS client apps.

Also, the idea of restructuring the backends concept to use a common character-I/O-based API makes sense to me. The challenge I see there is to make it intuitive and easy to use, but I'm willing to take part in that effort, should it get some traction.

kartben commented on samples/subsys/shell/bluetooth/README.rst:
What's with the empty file here? This really needs instructions on how one is supposed to use/test this :)
Agreed with another comment saying that having a script to exercise this would be useful, especially since the "client" FW looks a bit intimidating (and not a big fan of client/server terminology in this context)

ubieda replied:
Agreed with another comment saying that having a script to exercise this would be useful

Alright - I'll add this item on the TODO list capture on the description of the PR to get it done!

ubieda replied:
@kartben I addressed the following changes:
- Add README to the sample.
- Remove the user-side FW.
- Here's an app you can use to exercise the latest version of the Sample.
https://github.com/ubieda/bleak/blob/feature/interactive-terminal-over-bt/examples/uart_service.py

kartben commented on samples/subsys/shell/bluetooth/server/prj.conf:
```diff
@@ -0,0 +1,31 @@
+CONFIG_PRINTK=y
+CONFIG_SHELL=y
+CONFIG_LOG=y
+CONFIG_LOG_CMDS=y
+CONFIG_INIT_STACKS=y
+CONFIG_THREAD_STACK_INFO=y
+CONFIG_KERNEL_SHELL=y
+CONFIG_THREAD_MONITOR=y
+CONFIG_BOOT_BANNER=n
+CONFIG_THREAD_NAME=y
+CONFIG_DEVICE_SHELL=y
+CONFIG_POSIX_CLOCK=y
+CONFIG_DATE_SHELL=y
+CONFIG_THREAD_RUNTIME_STATS=y
+CONFIG_THREAD_RUNTIME_STATS_USE_TIMING_FUNCTIONS=y
+CONFIG_STATS=y
+CONFIG_STATS_SHELL=y
+CONFIG_SHELL_LOG_BACKEND=n
+
+CONFIG_BT=y
+CONFIG_BT_PERIPHERAL=y
+CONFIG_SHELL_BACKEND_BT=y
+CONFIG_BT_DEVICE_NAME="Shell-Sample"
+CONFIG_BT_RX_STACK_SIZE=2048
+# Allow for large Bluetooth data packets.
+CONFIG_BT_L2CAP_TX_MTU=512
+CONFIG_BT_BUF_ACL_RX_SIZE=502
+CONFIG_BT_BUF_ACL_TX_SIZE=502
+CONFIG_BT_CTLR_DATA_LENGTH_MAX=251
+
+CONFIG_ASSERT=y
```
this feels like this could use some cleanup? For the sake of keeping the sample as "minimal" as possible, please make sure there isn't more in there than what is required for demonstrating the functionality.

ubieda replied:
I'll look into it!

ubieda replied:
I re-organized the prj.conf to make it more readable. Most of those config's are just for the sake of having commands to demo. I can further simplify it if it need be.

kartben reviewed (changes requested):
the sample will need an actual an actual README please :)

ubieda commented:
@kartben FYI - I'm shooting for addressing all the pending action items of this PR before the week ends. I'll keep you posted!

jori-nordic commented:
> Have you asked Nordic's permissions to use this?

AFAIK this is not a registered 16-bit service, anyone is free to re-use 128-bit uuids. We can call it NUS-compatible if that's what's bugging you.

> This is duplicate of MCUmgr's shell functionality

Why don't we deprecate the Zephyr shell then in favor of that? If you read [my reply to Luis above](https://github.com/zephyrproject-rtos/zephyr/pull/68076#issuecomment-1926390515), you'll see that I'm definitely in favor of settling on a common character device API.
IMO this is a good stopgap, and we can deprecate all the bespoke log and shell transports when someone takes on that work.

ubieda commented:
@nordicjm thanks for your feedback!

I also appreciate having provided some guidance on understanding the implementation you’re suggesting already serves this purpose. Because of that, I was able to rapidly build the MCUMgr sample: smp_svr, configuring MCUMgr’s BT backend and enabling MCUMgr’s Shell commands handler.

As a result, I acknowledge that it is possible to with this implementation one is able to exchange shell commands over BLE as it is.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/65253049/331de78d-9bbb-462c-8120-a412325c5ca0)

> This is duplicate of MCUmgr's shell functionality (which supports UART, bluetooth, UDP and custom transports),

However, I disagree on the statement of this PR being a duplicate of the existing implmentation, due to the following reasons:

**1. Lack of documentation:** by looking at the README sample or even the Shell doc, it isn’t clearly demonstrated that this configuration is available for developers to use.
**2. Obfuscated/Not-clearly-accessible:** the availability of the feature seems somewhat hidden. The fact that it relies on SHELL_BACKEND_DUMMY, hints it may just be a byproduct/add-on of what’s already implemented on MCUMgr, and not a top-tier feature.
**3. Coupling with MCUMgr:** the feature does not stand on its own, because we need MCUMgr enabled for this to work. IMO, it should be  should be decoupled with any other subsystems, including MCUMgr.

The implementation of this PR is straightforward, decoupled and consistent with the usage of other Shell backends.

I do agree that we don’t want to maintain duplicate backends, and as @jori-nordic states, this exposes the need to refactor towards using a common-transport layer for the different subsystems, where they’re inter-dependent, clearly documented and consistent. Again, I’m willing to assist in that effort, wherever it makes sense.

As a result: This discussion could warrant the creation of a separate issue describing these needs to start thinking about that. Would you agree, @jori-nordic  @nordicjm?

> Have you asked Nordic's permissions to use this?


Provided we'd only be using the NUS UUIDs and not the underlying implementation, I'm not sure we're breaking the licensing terms by Nordic. In any case, I've already contacted Nordic to get some formal feedback.

thedjnK commented:
> 1. Lack of documentation: by looking at the README sample or even the Shell doc, it isn’t clearly demonstrated that this configuration is available for developers to use.

It's fully documented https://docs.zephyrproject.org/latest/services/device_mgmt/smp_groups/smp_group_9.html

>  2. Obfuscated/Not-clearly-accessible: the availability of the feature seems somewhat hidden. The fact that it relies on SHELL_BACKEND_DUMMY, hints it may just be a byproduct/add-on of what’s already implemented on MCUMgr, and not a top-tier feature.

Obfuscated? What? It need shell backend dummy because that's how it runs commands, that's not hidden.

>  3. Coupling with MCUMgr: the feature does not stand on its own, because we need MCUMgr enabled for this to work. IMO, it should be should be decoupled with any other subsystems, including MCUMgr.

This point is a bit ridiculous, MCUmgr is a transport, it allows you to send e.g. an 8KiB message over BLE, any form of using BLE (or other transport) needs this

> Why don't we deprecate the Zephyr shell then in favor of that? If you read [my reply to Luis above](https://github.com/zephyrproject-rtos/zephyr/pull/68076#issuecomment-1926390515), you'll see that I'm definitely in favor of settling on a common character device API.
> IMO this PR is a good stopgap, and we can deprecate all the bespoke log and shell transports when someone takes on that work.

@nordic-krch would need to comment on that

jori-nordic commented:
> As a result: This discussion could warrant the creation of a separate issue describing these needs to start thinking about that. Would you agree, @jori-nordic @nordicjm?

Definitely.

ubieda commented:
@thedjnK @nordicjm thanks for following-up with your reply.

Without addressing each one of your replies, I'll just say it's OK to disagree, while still respecting each others ideas.

carlescufi commented:
@ubieda thanks very much for this PR. I think the controversy around the duplication of functionality needs to be addressed before we can actually decide one way or the other. In parallel I will consult internally regarding the use of the UUID. I will add this PR to the Architecture WG so we can decide.

jori-nordic commented:
Regarding obfuscation, I'd like to add my sentiment. Please don't take this as flamebait, I'm just stating my experience/opinion developing end-user apps on Zephyr.

Zephyr has a whole lot of features and nifty build-related aids, which is a nice thing.
Unfortunately, it has IMO a discovery and documentation problem. A lot of things (as you say @thedjnK) are "fully documented". Some, less so.

One problem is finding this documentation. It is not easy if you're not a Zephyr Expert <sup>TM</sup>. That is if you haven't suffered slogging through the source code to find and understand the one feature you're trying to make work.

The other problem is maintainers forgetting about that initial pain, and just assuming everyone has the same knowledge of Zephyr and its intricacies. They then are surprised people don't know how to use this or that feature, as it is trivial to them, and also trivial for them to navigate the Zephyr docs.

A good example of that is writing a device driver. What takes less than an afternoon with a beginner-friendly ecosystem like Arduino can take the better part of a week with Zephyr. 
A developer has to:
- learn Kconfig, a weird configuration language (that can't do math for some reason)
- learn the Zephyr build system: yes, it's not standard cmake. You can't just export and link libraries.
- learn devicetree configuration (overlays etc), and its yaml bindings
- learn how to use DTS macros. Mostly by reading other drivers and trying stuff. You can't tell me with a straight face that the errors the compiler spits out are "good UX"
- learn how west projects work, where to put the magic files for zephyr to detect your new driver

Nordic has a few tutorials that demystify some of these concepts, but they are just scratching the surface.

I think it would really benefit the Zephyr project to have both reference documentation (what we currently have) *and* applied documentation, as in: I want to achieve this with Zephyr, here is exactly how we are going to do it. 

Bonus with that kind of documentation is you can explain away all the papercuts, or options that might make no sense at first sight (ie the DUMMY kconfig).
Having sample repos for out of tree drivers and applications is nice (as reference), but I think "showing your work" like in math class is better than just showing the answer.

NB:
Obfuscation is sometimes also in the form of badly named kconfig options or API that doesn't make sense to newcomers. I agree that relying on a kconfig with `DUMMY` in its name is suboptimal. The insanity of the BLE advertising options also comes to mind.

jori-nordic commented:
As per the WG discussion, this PR will not be further developed.

ubieda commented:
> Bluetooth WG:
> 
> * Discoverability of the functionality for users is very important
> * The argument about duplication would apply to physical UARTs too: Shell has both a UART transport and an MCUmgr transport (which in turn uses UART)
> * Agreement to implement NUS as a Bluetooth service, mapping it to the UART API
> * Agreement to use both the logging and shell UART backends instead of BLE-specific ones
>   
>   * CC @vChavezB your code from [logging: Added BLE Backend Service #56307](https://github.com/zephyrproject-rtos/zephyr/pull/56307) would be effectively removed from the tree
> * Agreement to create a snippet that represents a UART over BLE
> * Agreement to talk internally about formalizing the NUS spec
> * NUS generic service code should be in `subsys/bluetooth/services/nus.c`
> * UART driver code can either live in `subsys/bluetooth/services/nus.c` (just like USB CDC ACM) or `drivers/serial/nus.c`

Just to ellaborate on what was discussed in the Bluetooth WG:

- The claim of "duplication" omits the fact that commands over MCUmgr are encoded, which indeed implies that the existing app-integrators (PC, Mobile, etc) need to have an additional level of decoding (recognize it's an MCUmgr command, and that is the Shell command) before actually parsing the command. [Code Reference here](https://github.com/zephyrproject-rtos/zephyr/blob/1e974adb3331aef5f8eaa34ebcae2471387a5978/subsys/mgmt/mcumgr/grp/shell_mgmt/src/shell_mgmt.c#L60). I believe **there's no functional duplication, but rather functional similarities**.
- As a result, we agree there is value in making a Shell over Bluetooth that is MCUmgr-agnostic (regardless of being Backend-centric or abstracted behind UART) and that it must be clearly documented for users to make use of.
- WRT the observation of the "duplication" claim being also applicable to Shell UART Backend and MCUmgr UART backends: I don't have a historical context on when things were incorporated in the repo, but I would expect consistency when evaluating what is upstreamed vs incoming contributions. I would appreciate a clarification as to why it does not makes sense for BT but it does for others (e.g: UART).

@jori-nordic @carlescufi I'll follow-up with a separate PR with the agreed UART-centric implementation of Bluetooth, applicable for both Shell over Bluetooth and Logging over Bluetooth.

CC: @thedjnK / @nordicjm.

nordicjm commented:
> As a result, we agree there is value in making a Shell over Bluetooth that is MCUmgr-agnostic (regardless of being Backend-centric or abstracted behind UART) and that it must be clearly documented for users to make use of.

Not agreed, no, it needs to be brought up at the architecture working group to decide what the community as a whole agree on regarding duplication/acceptance

ubieda commented:
> > centric or abstracted behind UART) and that it must be clearly documented for users to make use of.
> 
> Not agreed, no, it needs to be brought up at the architecture working group to decide what the community as a whole agree on regarding duplication/acceptance

@nordicjm I had misunderstood it was something you, along with @jori-nordic, had agreed upon. I'm fine with discussing it in the next architecture WG meeting.

In the meanwhile, could you follow-up or ACK on the points 1 and 3? I'd like to reach a common-ground of agreement as to what we consider "duplicate" vs "similar", and define a consistent criteria to go by the current code structure, so we can get aligned in the end-goal.

nordicjm commented:
1. I don't really see the relevance here, both ends of any system need to have an understanding of the protocol, one could say that if you just throw data blinding without any sort of header to it then expect there to be problems with missing packets, what would this system do if a command was 80 bytes in length and the middle 20 bytes were just missing? It is functional duplication
2. 
3. MCUmgr UART backend is a binary protocol, it is not a shell

carlescufi commented:
> > As a result, we agree there is value in making a Shell over Bluetooth that is MCUmgr-agnostic (regardless of being Backend-centric or abstracted behind UART) and that it must be clearly documented for users to make use of.
> 
> Not agreed, no, it needs to be brought up at the architecture working group to decide what the community as a whole agree on regarding duplication/acceptance

@ubieda I just discussed this internally with @nordicjm and he has no objections to the proposal outlined in [this comment](https://github.com/zephyrproject-rtos/zephyr/pull/68076#issuecomment-1945971326). Another thing is additional discussions that may take place.

carlescufi commented:
> @jori-nordic @carlescufi I'll follow-up with a separate PR with the agreed UART-centric implementation of Bluetooth, applicable for both Shell over Bluetooth and Logging over Bluetooth.

Yes, please do, since this is, as I mentioned above, agreed by all.

ubieda commented:
> > > As a result, we agree there is value in making a Shell over Bluetooth that is MCUmgr-agnostic (regardless of being Backend-centric or abstracted behind UART) and that it must be clearly documented for users to make use of.
> > 
> > 
> > Not agreed, no, it needs to be brought up at the architecture working group to decide what the community as a whole agree on regarding duplication/acceptance
> 
> @ubieda I just discussed this internally with @nordicjm and he has no objections to the proposal outlined in [this comment](https://github.com/zephyrproject-rtos/zephyr/pull/68076#issuecomment-1945971326). Another thing is additional discussions that may take place.

@carlescufi @nordicjm what I want to make sure is if by "no objections" we mean the same "duplication" argument won't be brought up again. IMO - regardless whether we do it over UART or directly as a Shell Backend, this assumes we have a common-agreement on seeing the value in another Shell over BT implementation that does not depend in MCUmgr, as all the other Shell Backends currently do (including UART).

nordicjm commented:
> > > > As a result, we agree there is value in making a Shell over Bluetooth that is MCUmgr-agnostic (regardless of being Backend-centric or abstracted behind UART) and that it must be clearly documented for users to make use of.
> > > 
> > > 
> > > Not agreed, no, it needs to be brought up at the architecture working group to decide what the community as a whole agree on regarding duplication/acceptance
> > 
> > 
> > @ubieda I just discussed this internally with @nordicjm and he has no objections to the proposal outlined in [this comment](https://github.com/zephyrproject-rtos/zephyr/pull/68076#issuecomment-1945971326). Another thing is additional discussions that may take place.
> 
> @carlescufi @nordicjm what I want to make sure is if by "no objections" we mean the same "duplication" argument won't be brought up again. IMO - regardless whether we do it over UART or directly as a Shell Backend, this assumes we have a common-agreement on seeing the value in another Shell over BT implementation that does not depend in MCUmgr, as all the other Shell Backends currently do (including UART).

Bring it up at the architecture working group and discuss it there

ubieda commented:
Ok, may we please discuss it in the next Architecture WG meeting, @carlescufi?

ubieda commented:
> 1. I don't really see the relevance here, both ends of any system need to have an understanding of the protocol, one could say that if you just throw data blinding without any sort of header to it then expect there to be problems with missing packets, what would this system do if a command was 80 bytes in length and the middle 20 bytes were just missing? It is functional duplication

According to this argument: **Anything** that uses Shell and it's piped over BLE is a duplicate. I disagree with this statement (see below how it's applied to over transports), but we can discuss it in the meeting and see what others think.

> 3. MCUmgr UART backend is a binary protocol, it is not a shell

Please, consider:

**Scenario 1 - Duplicate claim:**
1. Shell -> MCUmgr -> BLE.
2. Shell -> BLE.

**Scenario 2 - The same duplicate claim applied on existing upstreamed code:**
1. Shell -> MCUmgr -> UART.
3. Shell -> UART.

carlescufi commented:
> Ok, may we please discuss it in the next Architecture WG meeting, @carlescufi?

OK, @ubieda. I will bring it up.
